{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0",
   "metadata": {},
   "outputs": [],
   "source": [
    "pip install -r /Users/mastorga/Documents/BTE-LLM/requirements.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "pip install https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/releases/v0.5.4/en_core_sci_sm-0.5.4.tar.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain_openai import ChatOpenAI\n",
    "from scispacy.linking import EntityLinker\n",
    "from pydantic import BaseModel, Field\n",
    "from langchain_core.tools import tool\n",
    "from typing import Annotated\n",
    "import re\n",
    "import spacy\n",
    "import requests\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "from nltk.corpus import stopwords\n",
    "from nltk.tokenize import word_tokenize"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# loading scispacy\n",
    "nlp = spacy.load(\"en_core_sci_sm\")\n",
    "nlp.add_pipe(\"scispacy_linker\", config={\"resolve_abbreviations\": True, \"linker_name\": \"umls\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def BioNERToolLLM(query: str):\n",
    "    global nlp\n",
    "    # getting linker\n",
    "    linker = nlp.get_pipe(\"scispacy_linker\")\n",
    "\n",
    "    llm = ChatOpenAI(temperature=0, model=\"gpt-4o\")\n",
    "    \n",
    "    bioIDs = {}\n",
    "    idList = {}\n",
    "\n",
    "    ent_prompt = f\"\"\"You are a helpful assistant that can extract biological terms/entities from a given query. \n",
    "                    These might include diseases (such as malaria, early onset dementia), genes, proteins, biological entities (such as viruses), etc.\n",
    "\n",
    "                    Do NOT include common nouns/phrases such as \"gene\", \"compound\", \"pathogen\", \"associated with\", etc. \n",
    "                    Your response must ONLY include discrete proper noun entities such as \"early onset dementia\" or \"sickle cell anemia\". \n",
    "                    Make sure to include symptoms such as \"fever\" or \"cough\".\n",
    "\n",
    "                    If extracting terms concerning proteins/genes such as \"ESR1 upregulation\", only return the gene/protein name (\"ESR1\").\n",
    "                    You must always return the full phrase/long form of each biomedical entity \n",
    "                    (for example, \"type I and type II diabetes\" should result in \"type I diabetes\" and \"type II diabetes\"; \"DNA polymerases of human vs. mice\" should result in \"human DNA polymerase\" and \"mouse DNA polymerase\")\n",
    "                    Return results as a list.\n",
    "                    Here is your query: {query}\"\"\"\n",
    "\n",
    "    entities = llm.invoke(ent_prompt).content\n",
    "\n",
    "    verify_prompt = f\"\"\"Given the following query, have all relevant biomedical terms been extracted?\n",
    "                    {query}\n",
    "\n",
    "                    Here are the extracted biomedical terms: {entities}\n",
    "                    \n",
    "                    Return TRUE if all biomedical entities were extracted. Return FALSE if there are any missing biomedical entities from the list.\n",
    "                    Your response MUST be boolean\"\"\"\n",
    "    \n",
    "    verified = llm.invoke(verify_prompt).content\n",
    "    \n",
    "    if verified.lower() != \"true\":\n",
    "        retry_prompt = f\"\"\"Given the following query, which biomedical terms are missing from the list?\n",
    "                    {query}\n",
    "\n",
    "                    Here are the extracted biomedical terms: {entities}\n",
    "\n",
    "                    Your response must ONLY include discrete proper noun entities such as \"early onset dementia\" or \"sickle cell anemia\". \n",
    "                    Make sure to include symptoms such as \"fever\" or \"cough\".\n",
    "                    \n",
    "                    Return a list of ONLY the missing biomedical terms\"\"\"\n",
    "        \n",
    "        addtl_ents = llm.invoke(retry_prompt).content\n",
    "        entities = entities + \" \" + addtl_ents\n",
    "    \n",
    "    print(entities)\n",
    "    \n",
    "    doc = nlp(entities)\n",
    "\n",
    "    for ent in doc.ents:\n",
    "        if ent._.kb_ents:  # Check if entity has linked knowledge base IDs\n",
    "            for id in ent._.kb_ents:\n",
    "                idList[id[0]] = remove_after_character(str(linker.kb.cui_to_entity[id[0]]), \"TUI\")\n",
    "                \n",
    "                print(idList[id[0]])\n",
    "\n",
    "                print('\\n')\n",
    "\n",
    "                print(getCosineSimilarity(query, idList[id[0]]))\n",
    "\n",
    "                print(\"\\n\\n\")\n",
    "\n",
    "            select_prompt = f\"\"\"You are a smart biomedical assistant that can understand the context and the intent behind a query. \n",
    "                        Be careful when choosing IDs for entities that can refer to different concepts (for example, HIV can refer either to the virus or the disease; you MUST choose the most appropriate concept/definition based on the query). \n",
    "                        Use the context and the intent behind the query to choose the most appropriate ID. \n",
    "                        Here is the complete query: {query}\n",
    "                        Select the one most appropriate ID/CUI for {ent.text} from the list below:\n",
    "                        {idList}\n",
    "                        If none of the choices are appropriate, return \"\".\n",
    "                        Otherwise, return only the ID/CUI.\n",
    "                        \"\"\"\n",
    "\n",
    "            # LLM selects most appropriate ID from list\n",
    "            selectedID = llm.invoke(select_prompt).content.strip()\n",
    "\n",
    "            # Extract just the UMLS CUI using regex\n",
    "            match = re.search(r\"C\\d{7}\", selectedID)\n",
    "            if match:\n",
    "                bioIDs[ent.text] = \"UMLS:\" + match.group(0)\n",
    "            else:\n",
    "                bioIDs[ent.text] = \"\"\n",
    "\n",
    "            print(ent.text + \" - \" + bioIDs[ent.text])\n",
    "\n",
    "    return bioIDs if bioIDs else {\"message\": \"No entities found\"}    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def BioNERToolcosine(query: str):\n",
    "    global nlp\n",
    "    # getting linker\n",
    "    linker = nlp.get_pipe(\"scispacy_linker\")\n",
    "\n",
    "    llm = ChatOpenAI(temperature=0, model=\"gpt-4o\")\n",
    "    \n",
    "    bioIDs = {}\n",
    "    idList = {}\n",
    "\n",
    "    ent_prompt = f\"\"\"You are a helpful assistant that can extract biological terms/entities from a given query. \n",
    "                    These might include diseases (such as malaria, early onset dementia, parkinsonism), genes, proteins, biological entities (such as viruses), etc.\n",
    "\n",
    "                    Do NOT include common nouns/phrases such as \"gene\", \"compound\", \"pathogen\", \"associated with\", etc. \n",
    "                    Your response must ONLY include discrete proper noun entities such as \"early onset dementia\" or \"sickle cell anemia\". \n",
    "                    Make sure to include symptoms such as \"fever\" or \"cough\".\n",
    "\n",
    "                    If extracting terms concerning proteins/genes such as \"ESR1 upregulation\", only return the gene/protein name (\"ESR1\").\n",
    "                    You must always return the full phrase/long form of each biomedical entity \n",
    "                    (for example, \"type I and type II diabetes\" should result in \"type I diabetes\" and \"type II diabetes\"; \"DNA polymerases of human vs. mice\" should result in \"human DNA polymerase\" and \"mouse DNA polymerase\")\n",
    "                    Return results as a list.\n",
    "                    Here is your query: {query}\"\"\"\n",
    "\n",
    "    entities = llm.invoke(ent_prompt).content\n",
    "\n",
    "    verify_prompt = f\"\"\"Given the following query, have all relevant biomedical terms been extracted?\n",
    "                    {query}\n",
    "\n",
    "                    Here are the extracted biomedical terms: {entities}\n",
    "                    \n",
    "                    Return TRUE if all biomedical entities were extracted. Return FALSE if there are any missing biomedical entities from the list.\n",
    "                    Your response MUST be boolean\"\"\"\n",
    "    \n",
    "    verified = llm.invoke(verify_prompt).content\n",
    "    \n",
    "    if verified.lower() != \"true\":\n",
    "        retry_prompt = f\"\"\"Given the following query, which biomedical terms are missing from the list?\n",
    "                    {query}\n",
    "\n",
    "                    Here are the extracted biomedical terms: {entities}\n",
    "\n",
    "                    Here were the extraction instructions: {ent_prompt}\n",
    "                    \n",
    "                    Return a list of ONLY the missing biomedical terms\"\"\"\n",
    "        \n",
    "        addtl_ents = llm.invoke(retry_prompt).content\n",
    "        entities = entities + \" \" + addtl_ents\n",
    "    \n",
    "    print(entities)\n",
    "    \n",
    "    doc = nlp(entities)\n",
    "\n",
    "    for ent in doc.ents:\n",
    "        if ent._.kb_ents:  # Check if entity has linked knowledge base IDs\n",
    "            for id in ent._.kb_ents:\n",
    "                idList[id[0]] = remove_after_character(str(linker.kb.cui_to_entity[id[0]]), \"TUI\")\n",
    "                \n",
    "                print(idList[id[0]])\n",
    "\n",
    "                print('\\n')\n",
    "\n",
    "                print(getCosineSimilarity(query, idList[id[0]]))\n",
    "\n",
    "                print(\"\\n\\n\")\n",
    "\n",
    "                print(str(linker.kb.cui_to_entity[id[0]]).get(\n",
    "\n",
    "            select_prompt = f\"\"\"You are a smart biomedical assistant that can understand the context and the intent behind a query. \n",
    "                        Be careful when choosing IDs for entities that can refer to different concepts (for example, HIV can refer either to the virus or the disease; you MUST choose the most appropriate concept/definition based on the query). \n",
    "                        Use the context and the intent behind the query to choose the most appropriate ID. \n",
    "                        Here is the complete query: {query}\n",
    "                        Select the one most appropriate ID/CUI for {ent.text} from the list below:\n",
    "                        {idList}\n",
    "                        If none of the choices are appropriate, return \"\".\n",
    "                        Otherwise, return only the ID/CUI.\n",
    "                        \"\"\"\n",
    "\n",
    "            # LLM selects most appropriate ID from list\n",
    "            selectedID = llm.invoke(select_prompt).content.strip()\n",
    "\n",
    "            # Extract just the UMLS CUI using regex\n",
    "            match = re.search(r\"C\\d{7}\", selectedID)\n",
    "            if match:\n",
    "                bioIDs[ent.text] = \"UMLS:\" + match.group(0)\n",
    "            else:\n",
    "                bioIDs[ent.text] = \"\"\n",
    "\n",
    "            print(ent.text + \" - \" + bioIDs[ent.text])\n",
    "\n",
    "    return bioIDs if bioIDs else {\"message\": \"No entities found\"}    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def BioNERTool(query: str):\n",
    "    \"\"\"Extract biological entities from a query and returns them along with their ID\"\"\"\n",
    "\n",
    "    def remove_TUI(text):\n",
    "        parts = text.split(\"TUI\", 1)\n",
    "        return parts[0]\n",
    "\n",
    "    # Setting up nlp model\n",
    "    global nlp\n",
    "    linker = nlp.get_pipe(\"scispacy_linker\")\n",
    "\n",
    "    bioIDs = {}\n",
    "    idList = {}\n",
    "\n",
    "    llm = ChatOpenAI(temperature=0, model=\"gpt-4o\")\n",
    "\n",
    "    ent_prompt = f\"\"\"You are a helpful assistant that can extract biological terms/entities from a given query. \n",
    "                    These might include diseases (such as malaria, early onset dementia, parkinsonism), genes, proteins, biological entities (such as viruses), etc.\n",
    "\n",
    "                    Do NOT include common nouns/phrases such as \"gene\", \"compound\", \"pathogen\", \"associated with\", etc. \n",
    "                    Your response must ONLY include discrete proper noun entities such as \"early onset dementia\" or \"sickle cell anemia\". \n",
    "                    Make sure to include symptoms such as \"fever\" or \"cough\".\n",
    "\n",
    "                    If extracting terms concerning proteins/genes such as \"ESR1 upregulation\", only return the gene/protein name (\"ESR1\").\n",
    "                    You must always return the full phrase/long form of each biomedical entity \n",
    "                    (for example, \"type I and type II diabetes\" should result in \"type I diabetes\" and \"type II diabetes\"; \"DNA polymerases of human vs. mice\" should result in \"human DNA polymerase\" and \"mouse DNA polymerase\")\n",
    "                    Return results as a list.\n",
    "                    Here is your query: {query}\"\"\"\n",
    "\n",
    "    entities = llm.invoke(ent_prompt).content\n",
    "\n",
    "    verify_prompt = f\"\"\"Given the following query, have all relevant biomedical terms been extracted?\n",
    "                    {query}\n",
    "\n",
    "                    Here are the extracted biomedical terms: {entities}\n",
    "                    \n",
    "                    Return TRUE if all biomedical entities were extracted. Return FALSE if there are any missing biomedical entities from the list.\n",
    "                    Your response MUST be boolean\"\"\"\n",
    "    \n",
    "    verified = llm.invoke(verify_prompt).content\n",
    "    \n",
    "    if verified.lower() != \"true\":\n",
    "        retry_prompt = f\"\"\"Given the following query, which biomedical terms are missing from the list?\n",
    "                    {query}\n",
    "\n",
    "                    Here are the extracted biomedical terms: {entities}\n",
    "\n",
    "                    Here were the extraction instructions: {ent_prompt}\n",
    "                    \n",
    "                    Return a list of ONLY the missing biomedical terms\"\"\"\n",
    "        \n",
    "        addtl_ents = llm.invoke(retry_prompt).content\n",
    "        entities = entities + \" \" + addtl_ents\n",
    "    \n",
    "    print(entities + \"\\n\")\n",
    "    \n",
    "    doc = nlp(entities)\n",
    "\n",
    "    for ent in doc.ents:\n",
    "        if ent._.kb_ents:  # Check if entity has linked knowledge base IDs\n",
    "            for id in ent._.kb_ents:\n",
    "                idList[id[0]] = remove_TUI(str(linker.kb.cui_to_entity[id[0]]))\n",
    "\n",
    "            select_prompt = f\"\"\"You are a smart biomedical assistant that can understand the context and the intent behind a query. \n",
    "                        Be careful when choosing IDs for entities that can refer to different concepts (for example, HIV can refer either to the virus or the disease; you MUST choose the most appropriate concept/definition based on the query). \n",
    "                        Use the context and the intent behind the query to choose the most appropriate ID. \n",
    "                        Here is the complete query: {query}\n",
    "                        Select the one most appropriate ID/CUI for {ent.text} from the list below:\n",
    "                        {idList}\n",
    "                        If none of the choices are appropriate, return \"\".\n",
    "                        Otherwise, return only the ID/CUI.\n",
    "                        \"\"\"\n",
    "\n",
    "            # LLM selects most appropriate ID from list\n",
    "            selectedID = llm.invoke(select_prompt).content.strip()\n",
    "\n",
    "            # Extract just the UMLS CUI using regex\n",
    "            match = re.search(r\"C\\d{7}\", selectedID)\n",
    "            if match:\n",
    "                bioIDs[ent.text] = \"UMLS:\" + match.group(0)\n",
    "                definition = idList[match.group(0)]\n",
    "            else:\n",
    "                bioIDs[ent.text] = \"\"\n",
    "\n",
    "            print(ent.text + \" - \" + bioIDs[ent.text] + '\\n' + definition)\n",
    "\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getCosineSimilarity(str1: str, str2: str):\n",
    "    # Tokenizing strings\n",
    "    str1_list = word_tokenize(str1)\n",
    "    str2_list = word_tokenize(str2)\n",
    "\n",
    "    # sw contains the list of stopwords\n",
    "    sw = stopwords.words('english')\n",
    "    l1 = [];l2 = []\n",
    "\n",
    "    # Removing stop words from the string\n",
    "    str1_set = {w for w in str1_list if not w in sw}\n",
    "    str2_set = {w for w in str2_list if not w in sw}\n",
    "\n",
    "    # Forming a set containing the keywords of both strings\n",
    "    rvector = str1_set.union(str2_set)\n",
    "    for w in rvector:\n",
    "        if w in str1_set: l1.append(1)\n",
    "        else: l1.append(0)\n",
    "        if w in str2_set: l2.append(1)\n",
    "        else: l2.append(0)\n",
    "\n",
    "    c = 0\n",
    "\n",
    "    # Cosine formula\n",
    "    for i in range(len(rvector)):\n",
    "        c += l1[i]*l2[i]\n",
    "\n",
    "    cosine = c / float((sum(l1)*sum(l2))**0.5)\n",
    "\n",
    "    return cosine"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9",
   "metadata": {},
   "outputs": [],
   "source": [
    "sampleent1 = \"\"\"CUI: C0019682, Name: HIV\n",
    "Definition: Human immunodeficiency virus. A non-taxonomic and historical term referring to any of two species, specifically HIV-1 and/or HIV-2. Prior to 1986, this was called human T-lymphotropic virus type III/lymphadenopathy-associated virus (HTLV-III/LAV). From 1986-1990, it was an official species called HIV. Since 1991, HIV was no longer considered an official species name; the two species were designated HIV-1 and HIV-2.\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10",
   "metadata": {},
   "outputs": [],
   "source": [
    "sampleent2 = \"\"\"CUI: C0019693, Name: HIV Infections\n",
    "Definition: Includes the spectrum of human immunodeficiency virus infections that range from asymptomatic seropositivity, thru AIDS-related complex (ARC), to acquired immunodeficiency syndrome (AIDS).\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11",
   "metadata": {},
   "outputs": [],
   "source": [
    "sampleent3 = \"\"\"CUI: C0019704, Name: HIV-1\n",
    "Definition: The type species of LENTIVIRUS and the etiologic agent of AIDS. It is characterized by its cytopathic effect and affinity for the T4-lymphocyte.\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12",
   "metadata": {},
   "outputs": [],
   "source": [
    "sampleent4 = \"\"\"CUI: C0019707, Name: Human immunodeficiency virus 2 (HIV-2)\n",
    "Definition: An HIV species related to HIV-1 but carrying different antigenic components and with differing nucleic acid composition. It shares serologic reactivity and sequence homology with the simian Lentivirus SIMIAN IMMUNODEFICIENCY VIRUS and infects only T4-lymphocytes expressing the CD4 phenotypic marker.\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13",
   "metadata": {},
   "outputs": [],
   "source": [
    "getCosineSimilarity(\"What receptor does the HIV virus bind to?\", sampleent1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14",
   "metadata": {},
   "outputs": [],
   "source": [
    "getCosineSimilarity(\"What receptor does the HIV virus bind to?\", sampleent2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15",
   "metadata": {},
   "outputs": [],
   "source": [
    "getCosineSimilarity(\"What receptor does the HIV virus bind to?\", sampleent3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16",
   "metadata": {},
   "outputs": [],
   "source": [
    "getCosineSimilarity(\"What receptor does the HIV virus bind to?\", sampleent4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17",
   "metadata": {},
   "outputs": [],
   "source": [
    "getCosineSimilarity(\"What drugs can treat HIV?\", sampleent1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18",
   "metadata": {},
   "outputs": [],
   "source": [
    "getCosineSimilarity(\"What drugs can treat HIV?\", sampleent2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19",
   "metadata": {},
   "outputs": [],
   "source": [
    "getCosineSimilarity(\"What drugs can treat HIV?\", sampleent3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20",
   "metadata": {},
   "outputs": [],
   "source": [
    "getCosineSimilarity(\"What drugs can treat HIV?\", sampleent4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21",
   "metadata": {},
   "outputs": [],
   "source": [
    "BioNERTool(\"What drugs can treat Parkinsonism?\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22",
   "metadata": {},
   "outputs": [],
   "source": [
    "BioNERToolLLM(\"What receptor does HIV bind to?\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23",
   "metadata": {},
   "outputs": [],
   "source": [
    "import nltk\n",
    "\n",
    "nltk.download('punkt_tab')\n",
    "nltk.download('stopwords')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24",
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_after_character(text, character):\n",
    "    \"\"\"\n",
    "    Removes all characters in a string after the first occurrence of a specified character.\n",
    "\n",
    "    Args:\n",
    "        text (str): The input string.\n",
    "        character (str): The character to split the string by.\n",
    "\n",
    "    Returns:\n",
    "        str: The modified string with everything after the character removed.\n",
    "             Returns the original string if the character is not found.\n",
    "    \"\"\"\n",
    "    parts = text.split(character, 1) # Split only at the first occurrence\n",
    "    return parts[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
